Added no-assign-parameter rule

Cisplatin opened the pull request:
As requested in https://augur.net/bounties/: an implementation of the rule 'Do not assign to parameters'.

Please let me know if there's anything I can change!

duaraghav8 commented on README.md:
```diff
@@ -42,6 +42,7 @@ Below are the rules supplied by this plugin and the information on passing optio
 | no-block-members               | Discourage use of members 'blockhash' & 'timestamp' (and alias 'now') of 'block' global variable | List of members to warn against   | ["blockhash", "timestamp"]           |       |
 | no-call-value                  | Discourage use of .call.value()()                                                                |                                   |                                      |       |
 | no-continue                    | Discourage use of 'continue' statement.                                                          |                                   |                                      |       |
+| no-assign-parameter            | Do not assign to parameters.                                                                     |                                   |                                      |       |
```
`Disallow assigning to function parameters.`
(Please make this change everywhere in this PR where description is given)

duaraghav8 commented on rules/no-assign-parameter.js:
```diff
@@ -0,0 +1,45 @@
+
+            var node = emitted.node;
+            let params = node.params.map(x => x['id']);
+            for (let statement of node.body.body) {
```
in case of a fallback function, `node.body.body` will throw exception. Please account for this + add tests
first ensure that `node.is_abstract` is `false`

duaraghav8 commented on test/no-assign-parameter.js:
```diff
@@ -0,0 +1,62 @@
+                done();
+        });
+
+        it('shouldn\'t raise an error for an assigned local variable', function(done) {
```
please add tests to see whether it also catches assignments that are inside `if..else` statements and all the loops.

duaraghav8 replied:
Also add tests for incrementing & decrementing integers

Cisplatin replied:
Good call! Added tests for those cases.

duaraghav8 commented on rules/no-assign-parameter.js:
```diff
@@ -0,0 +1,45 @@
+            var node = emitted.node;
+            let params = node.params.map(x => x['id']);
+            for (let statement of node.body.body) {
+                if(statement['expression']['type'] === 'AssignmentExpression' &&
```
If I understand correctly, you traverse through all the top-level stmts inside the function block. Couple of things:
- Every statement is not guaranteed to have the `expression` attribute. You should first ensure that the statement you're checking is an `ExpressionStatement`. Otherwise results in exception
- This technique only catches top-level assignments. If the assignment is happening inside a for loop or if/else block, I don't think it will catch it.

Please correct me if I'm wrong above^

Cisplatin commented:
@duaraghav8 Fixed the cases that you were talking about - great catches, missed those. Everything should be working properly now.

Cisplatin commented on rules/no-assign-parameter.js:
```diff
@@ -0,0 +1,75 @@
+                }
+
+                if (['ForStatement', 'WhileStatement', 'DoWhileStatement'].indexOf(statement['type']) >= 0) {
+                    inspectBody (statement['body']['body'], node, params);
```
This works, but I'm not sure if it would be a better approach to catch each of these declarations were caught separately from FunctionDeclarations. Any suggestions on that?

duaraghav8 replied:
Your technique is fine. Although it might not catch all issues, it assures that there will be no false positives. Please make sure it
- Catches any top-level assignments
- Catches any assignments happening in any of the loops
- catches inside if & else statements (they could be blocks or single statements)
eg
```
if (true) {...}

if (true)
    boo = 100;
```
You need to account for both

Also use a simple recursion trick: take the inspection of If & Loop out into their own functions. Inside these functions, just add a condition that if another `if` exists, call inspectIf or if loop exists, call inspectLoop (recursive).
This ensures that no matter how many levels deep that assignment resides, it will still get caught because the rule traverses every level

```
function foo(uint x) {
  if (true) {
    do {
      x += 1;
    } while (x < 10);
  }
}
```

Add tests for all above^ and we're good to merge!

Hope I was able to clearly explain it? If not, please let me know

duaraghav8 commented on rules/no-assign-parameter.js:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        function inspectFunctionDeclaration (emitted) {
+            if (emitted.exit) { return; }
```
please refactor to `if (emitted.exit || node.is_abstract) { return; }` and delete the below `if` statement

Cisplatin commented:
@duaraghav8 Good catch (again) with the false negatives - I've fixed the recursive solution so that it properly traverses the code. I also added the tests you mentioned to make sure that it works.

duaraghav8 commented on test/no-assign-parameter.js:
```diff
@@ -50,111 +50,29 @@ describe('[RULE] no-assign-parameter: Acceptances', function () {
+                             'contract Blah { function foo(uint abc) { while(true) abc = 12; } }',
+                             'contract Blah { function foo(uint abc) { while(true) while(true) abc = 12; } }',
+                             'contract Blah { function foo(uint x) { if (true) { do { x += 1; } while (x < 10); } } }'
+                            ]
```
semicolon after closing brace :P

duaraghav8 commented on rules/no-assign-parameter.js:
```diff
@@ -20,47 +20,65 @@ module.exports = {
+            if(isBadAssignment(statement, params) || isBadUpdate(statement, params)) {
+                context.report({
+                    node: node,
+                    message: 'Disallow assigning to function parameters'
```
Just to clarify, the message is different from description. The message is shown to the user. So here, the message should be `"<Function Name>": Avoid assigning to function parameters.`

duaraghav8 commented on rules/no-assign-parameter.js:
```diff
@@ -32,10 +32,11 @@ module.exports = {
         }
 
         function checkExpressionStatement(statement, node, params) {
+            let function_name = node.name ? `"${node.name}"` : 'Fallback function';
```
Could you please name it `functionName` instead since mixedCase is the convention throughout the repo

Cisplatin replied:
Of course - just changed that @duaraghav8
